pkg/wireguard/agent/agent.go

```diff
@@ -0,0 +1,356 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This package contains the agent code used to configure the Wireguard tunnel
+// between nodes. The code supports adding and removing peers at run-time
+// and the peer information is retrieved via the CiliumNode object.
+package agent
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"strconv"
+
+	"github.com/cilium/cilium/pkg/cidr"
+	"github.com/cilium/cilium/pkg/datapath/linux/linux_defaults"
+	"github.com/cilium/cilium/pkg/datapath/linux/route"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/sysctl"
+	"github.com/cilium/cilium/pkg/wireguard/types"
+
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+	"golang.zx2c4.com/wireguard/wgctrl"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+)
+
+const (
+	listenPort = 51871
+)
+
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "wireguard")
+
+type Agent struct {
+	lock.RWMutex
+	wgClient         *wgctrl.Client
+	listenPort       int
+	privKey          wgtypes.Key
+	wireguardV4CIDR  *net.IPNet
+	wireguardV6CIDR  *net.IPNet
+	pubKeyByNodeName map[string]string // nodeName => pubKey
+	restoredPubKeys  map[string]struct{}
+}
+
+func NewAgent(privKeyPath string, wgV4Net, wgV6Net *net.IPNet) (*Agent, error) {
+	key, err := loadOrGeneratePrivKey(privKeyPath)
+	if err != nil {
+		return nil, err
+	}
+
+	node.SetWireguardPubKey(key.PublicKey().String())
```

Comment:
Would it be better to do this after wgctrl.New() below in case the latter fails?